arithmetization/src/main/java/net/consensys/linea/zktracer/module/shakiradata/ShakiraDataOperation.java

```diff
@@ -0,0 +1,284 @@
+    return totalNumberOfDataRows() + RESULT_ROWS_COUNT + EXTRA_ROWS_COUNT;
+  }
+
+  private Map<Integer, PhaseInfo> getPhaseInfoMap() {
+    final int indexMaxDataPhases = totalNumberOfDataRows() - 1;
+    final int totalSizeDataPhases = hashInput.size();
+    final UnsignedByte[] deltaBytesDataPhases = getDataPhaseDeltaBytes();
+    final UnsignedByte[] deltaBytesResultPhases = getResultPhaseDeltaBytes();
+    final UnsignedByte[] deltaBytesExtraPhases = getExtraPhaseDeltaBytes();
+
+    return Map.of(
+        Trace.PHASE_SHA2_DATA,
+        new PhaseInfo(
+            Trace.PHASE_SHA2_DATA, indexMaxDataPhases, totalSizeDataPhases, deltaBytesDataPhases),
+        Trace.PHASE_SHA2_RESULT,
+        new PhaseInfo(
+            Trace.PHASE_SHA2_RESULT,
+            Trace.INDEX_MAX_SHA2_RESULT,
+            TOTAL_SIZE_RESULT_PHASES,
+            deltaBytesResultPhases),
+        Trace.PHASE_RIPEMD_DATA,
+        new PhaseInfo(
+            Trace.PHASE_RIPEMD_DATA, indexMaxDataPhases, totalSizeDataPhases, deltaBytesDataPhases),
+        Trace.PHASE_RIPEMD_RESULT,
+        new PhaseInfo(
+            Trace.PHASE_RIPEMD_RESULT,
+            Trace.INDEX_MAX_RIPEMD_RESULT,
+            TOTAL_SIZE_RESULT_PHASES,
+            deltaBytesResultPhases),
+        Trace.PHASE_KECCAK_DATA,
+        new PhaseInfo(
+            Trace.PHASE_KECCAK_DATA, indexMaxDataPhases, totalSizeDataPhases, deltaBytesDataPhases),
+        Trace.PHASE_KECCAK_RESULT,
+        new PhaseInfo(
+            Trace.PHASE_KECCAK_RESULT,
+            Trace.INDEX_MAX_KECCAK_RESULT,
+            TOTAL_SIZE_RESULT_PHASES,
+            deltaBytesResultPhases),
+        PHASE_EXTRA,
+        new PhaseInfo(
+            PHASE_EXTRA, INDEX_MAX_EXTRA, TOTAL_SIZE_EXTRA_PHASES, deltaBytesExtraPhases));
+  }
+
+  private Bytes extractLimbRowData(Integer rowIndex) {
+    final int inputDataSize = hashInput.size();
+    final int startLimbOffsetIndex = Trace.LLARGE * rowIndex;
+
+    final boolean limbSizeIsLessThan16 = inputDataSize - startLimbOffsetIndex < Trace.LLARGE;
+
+    return !limbSizeIsLessThan16
+        ? hashInput.slice(startLimbOffsetIndex, Trace.LLARGE)
+        : rightPadTo(
+            hashInput.slice(startLimbOffsetIndex, inputDataSize - startLimbOffsetIndex),
+            Trace.LLARGE);
+  }
+
+  private UnsignedByte[] getHubStampDiffBytes(int currentHubStamp) {
+    final BigInteger prevHubStampBigInt = BigInteger.valueOf(prevId);
+    final BigInteger hubStampBigInt = BigInteger.valueOf(currentHubStamp);
+    final BigInteger hubStampDiff =
+        hubStampBigInt.subtract(prevHubStampBigInt).subtract(BigInteger.ONE);
+
+    Preconditions.checkArgument(
+        hubStampDiff.compareTo(BigInteger.valueOf((long) Math.pow(256, 4))) < 0,
+        "Hub stamp difference should never exceed 256 ^ 4");
+
+    return bytesToUnsignedBytes(leftPadTo(bigIntegerToBytes(hubStampDiff), 4).toArray());
+  }
+
+  private UnsignedByte[] getDataPhaseDeltaBytes() {
```

Comment:
Why do we do all of this ? DEltaByte will be always 0 during data phase, no ?

Reply:
> DEltaByte will be always 0 during data phase, no ?

No, `deltaBytes` is an array of length `L` where
```java
L = ceil(size / 16) + 2 + 3
// ceil(size / 16) many data rows
// 2 result rows
// 3 'extra' rows
```
The contents of `deltaBytes` are all zero except (potentially) for the first 4 bytes and the last 2. The first byte is always on a data row.

**Note:** Since we deal with nonempty data `L >= 6` so there is always enough room. This is the purpose of the 'extra' rows.